### PR TITLE
Peer to peer and unsigned summary file support

### PR DESCRIPTION
**Pull request by pwithnall**
Here’s my branch which adds support for installing/updating apps/runtimes from LAN/USB, including the necessary support for unsigned summary files and the `ostree-metadata` branch (see https://github.com/ostreedev/ostree/issues/983).

It adds a new `--enable-p2p` option to hide the new functionality behind, but most of the commits have not been updated to actually conditionalise their changes on that option yet. Thus this branch is ready for initial review, but should not be merged yet. I’ll update it soon to add the `#ifdef`s, but they shouldn’t impact the overall behaviour or structure of the branch much.

There are also a bunch of `TODO` comments in the final commit which indicate the things I still need to check or implement (they will mostly be corner cases or upstreaming things to libostree; see below).

It adds a couple of new tests, but there are bound to be quite a few corner cases which I haven’t handled yet. I plan to test it for a while with `--enable-p2p` before we consider dropping the option.

### CODE REVIEW

File: common/flatpak-dir.c [hunk 1]
```diff
@@ -5674,6 +5685,40 @@ flatpak_dir_check_for_update (FlatpakDir          *self,
...
+      g_autoptr(GMainContext) context = NULL;
+      g_autoptr(GAsyncResult) find_result = NULL;
+      g_auto(OstreeRepoFinderResultv) results = NULL;
+      OstreeCollectionRef collection_ref = { collection_id, (char *) ref };
+      OstreeCollectionRef *collection_refs_to_fetch[2] = { &collection_ref, NULL };
+
+      context = g_main_context_new ();
+      g_main_context_push_thread_default (context);
+
+      ostree_repo_find_remotes_async (self->repo, (const OstreeCollectionRef * const *) collection_refs_to_fetch,
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: Surely you need some ifdefs here.

File: app/flatpak-builtins-build-export.c [hunk 2]
```diff
@@ -652,7 +653,7 @@ flatpak_builtin_build_export (int argc, char **argv, GCancellable *cancellable,
   guint64 installed_size = 0,download_size = 0;
   GTimeVal ts;
 
-  context = g_option_context_new (_("LOCATION DIRECTORY [BRANCH] - Create a repository from a build directory"));
+  context = g_option_context_new (_("LOCATION DIRECTORY [[COLLECTION-ID] BRANCH] - Create a repository from a build directory"));
```
Comment (alexlarsson) [inline app/flatpak-builtins-build-export.c]: Not sure about the order here. You might want to have a branch, but not set a collection id, whereas there is always a branchname, even if you leave the argument out.

In fact, isn't it better if this is --collection-id=ID

File: app/flatpak-builtins-build-export.c [hunk 3]
```diff
@@ -836,6 +866,9 @@ flatpak_builtin_build_export (int argc, char **argv, GCancellable *cancellable,
   if (!flatpak_repo_collect_sizes (repo, root, &installed_size, &download_size, cancellable, error))
     goto out;
 
+  g_variant_dict_insert_value (&metadata_dict, "ostree.commit.collection-ref",
+                               g_variant_new ("(ss)", (collection_id != NULL) ? collection_id : "", full_branch));
```
Comment (alexlarsson) [inline app/flatpak-builtins-build-export.c]: Shouldn't we inherit the current repo collection id if there is one rather than NULL?

**Comment by alexlarsson** [comment]
Don't we need a "collection id" property on FlatpakRemote in libflatpak.

File: common/flatpak-dir.c [hunk 4]
```diff
@@ -1929,6 +1932,7 @@ repo_pull_one_dir (OstreeRepo          *self,
       collection_refs_to_fetch[0] = &collection_ref;
       collection_refs_to_fetch[1] = NULL;
 
+      /* TODO: This implies the summary is downloaded elsewhere. Is this option actually accepted? */
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: Yes, we already did this when looking for updates. So we avoid ensure we get exactly the same version we detected then, as various decisions (like permissons) may have been done based on the metadata for that commit.

File: common/flatpak-dir.c [hunk 5]
```diff
@@ -8886,6 +8914,7 @@ flatpak_dir_update_remote_configuration (FlatpakDir   *self,
     return flatpak_dir_update_remote_configuration_for_repo_metadata (self, remote, summary, FALSE, NULL, cancellable, error);
 }
 
+/* TODO: Move these sizes out to per-ref or per-commit metadata */
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: This comment scares me a bit. We don't want to be in the situation where we have gnome-software blocking on startup until it has downloaded all the commit objects in all remotes before it starts showing anything, that will be very slow.

We already store this info in the commit metadata, but we propagate it to the summary for this reason. The summary file is a single, reasonlably compact file that we can download once and the have all info we need.

This, in combination with using unsigned summaries implies that we have to keep track of what the summary metadata said about a specific revision, and then verify that it was actually true when we pull the commit. I.e. we should not only verify that the signed commit has the same ref and collection id, but also that the metadata is identical to what was in the summary.

**Comment by alexlarsson** [comment]
So, i had a quick look at tihs code, and overall it looks reasonable.
I have two worries based on the comments and the code.

First of all (as i wrote in the comment) is that we want to continue relying on just downloading the summary file, and then have all the data we need for e.g command line completion, automatic dependency resolution (i.e. know what runtime and extensions are affected when installing/updating), etc. We don't want to actually download all the commits in the repo. 

Secondly, I worry a bit about all the mentionings of things being racy in the dynamic case. We do a search for revs at one point, surely we remember then also which rev came from where in the results we carry forward? We don't ask some randomly picked remote about some rev that some other remote said it had?

File: common/flatpak-dir.c (hunk 1 above)
Comment (pwithnall) [inline common/flatpak-dir.c]: Yeah, a lot of the necessary `#ifdef`s are missing in this WIP version of the branch.

File: app/flatpak-builtins-build-export.c (hunk 2 above)
Comment (pwithnall) [inline app/flatpak-builtins-build-export.c]: > Not sure about the order here. You might want to have a branch, but not set a collection id, whereas there is always a branchname, even if you leave the argument out.

Those cases can be differentiated by looking at `argc`, but I agree that it’s a bit confusing. I put the collection ID before the branch name because it’s always presented before the branch name elsewhere — i.e. like in the tuple `($collection_id, $ref_name)`.

> In fact, isn't it better if this is --collection-id=ID

I decided against that because ideally most flatpaks are (eventually) going to have a collection ID set, so it would ideally be a positional argument. However, `--collection-id` is probably a better solution to prevent the weirdness with `argc` and ordering. I’ll change it.

**Comment by pwithnall** [comment]
> Don't we need a "collection id" property on FlatpakRemote in libflatpak.

I did originally have a commit which added one, but then I realised the new API was never actually used, so I dropped the commit so that the API diff was minimised. I was thinking it could be added later once someone actually needs it.

(Adding public API which is ‘experimental’ turns into a large amount of pain in terms of advertising it in the pkg-config file, guarding it in the headers correctly for internal and potential external use, conditionally putting it in the documentation, etc.)

File: common/flatpak-dir.c (hunk 5 above)
Comment (pwithnall) [inline common/flatpak-dir.c]: > This comment scares me a bit. We don't want to be in the situation where we have gnome-software blocking on startup until it has downloaded all the commit objects in all remotes before it starts showing anything, that will be very slow.

The comment’s not very well worded, and it was more of a note to myself than to others. It’s to remind me to look into moving the entries from `xa.cache` into the additional metadata for each ref in the `summary` file (i.e. the `a{sv}` in the 0th child of `OSTREE_SUMMARY_GVARIANT_FORMAT`) so that the `xa.cache` property itself is split up and eliminated.

The bit about ‘per-commit metadata’ is just wrong.

**Comment by pwithnall** [comment]
> First of all (as i wrote in the comment) is that we want to continue relying on just downloading the summary file, and then have all the data we need for e.g command line completion, automatic dependency resolution (i.e. know what runtime and extensions are affected when installing/updating), etc. We don't want to actually download all the commits in the repo.

This branch does continue to rely on the metadata from the `summary` file (or, equivalently, the head commit in the `ostree-metadata` branch) and shouldn’t download all the commits in the repo. The comment which made you worry about this was poorly worded, sorry (see above).

> Secondly, I worry a bit about all the mentionings of things being racy in the dynamic case. We do a search for revs at one point, surely we remember then also which rev came from where in the results we carry forward? We don't ask some randomly picked remote about some rev that some other remote said it had?

The `OstreeRepoFinderResult **results` array does this for us — it specifies which refs are available from which peers, and the checksums for each of those refs. The comments you’re referring to predate when I modified `check_for_updates()` to propagate the `OstreeRepoFinderResult **results` array, which should do what we want to eliminate races. However, the comments are still partially valid, because there is some cleanup and verification I can do on a few code paths to check the results from `check_for_updates()` are actually used and there are actually no races. I’ll do that before I drop the comment.

**Comment by pwithnall** [comment]
I’ve just force-pushed an updated version of the branch, rebased on master and with quite a few fixes squashed in. It should now have all the necessary `#ifdef`s. There’s one niggle about conditional enum parsing for `glib-mkenums` which I need to solve at some point, but this is ready for detailed review.

@mwleeds, @matthiasclasen, would you mind taking a look at this? (Matthias, I don’t know if you’re taking over all code review while Alex is on holiday, but I thought I should ask anyway. :-) )

**Comment by matthiasclasen** [comment]
i'm trying to keep obvious things flowing. This one may not be obvious enough for me, but I can give it a look

File: lib/flatpak-remote.c [hunk 6]
```diff
@@ -160,6 +182,19 @@ flatpak_remote_class_init (FlatpakRemoteClass *klass)
                                                         "The name of the remote",
                                                         NULL,
                                                         G_PARAM_READWRITE));
+
+#ifdef FLATPAK_ENABLE_P2P
+#ifndef __GI_SCANNER__
+  g_object_class_install_property (object_class,
+                                   PROP_TYPE,
+                                   g_param_spec_enum ("type",
```
Comment (pwithnall) [inline lib/flatpak-remote.c]: Urgh, this is fallout from me marking the `FlatpakRemoteType` enum members as to be skipped, to avoid them getting picked up by `glib-mkenums` even when P2P support is disabled (because it ignores the `#ifdef FLATPAK_ENABLE_P2P` around them). I think the fix here is to unconditionally skip them, and manually add a `get_type()` function for them (conditionally).

File: common/flatpak-dir.c [hunk 7]
```diff
@@ -1839,6 +1857,29 @@ default_progress_changed (OstreeAsyncProgress *progress,
...
+ * %NULL. Return %TRUE if the ID was fetched successfully, or if it was unset or
+ * empty. */
+static gboolean
+repo_get_remote_collection_id (OstreeRepo  *repo,
+                               const char  *remote_name,
+                               char       **collection_id_out,
+                               GError     **error)
+{
+#ifdef FLATPAK_ENABLE_P2P
+  if (!ostree_repo_get_remote_option (repo, remote_name, "collection-id",
```
Comment (pwithnall) [inline common/flatpak-dir.c]: Yup.

File: common/flatpak-dir.c [hunk 8]
```diff
@@ -5658,21 +6082,27 @@ flatpak_dir_update (FlatpakDir          *self,
                                                       &gpg_verify_summary, error))
         return FALSE;
 
+      if (!repo_get_remote_collection_id (self->repo, remote_name, &collection_id, NULL))
+        collection_id = NULL;
```
Comment (pwithnall) [inline common/flatpak-dir.c]: Good catch. I was being relaxed about handling invalid configuration here, but there should be no harm in being strict.

File: common/flatpak-dir.c [hunk 9]
```diff
@@ -7480,8 +7941,10 @@ create_origin_remote_config (OstreeRepo   *repo,
   g_key_file_set_string (new_config, group, "xa.title", title);
   g_key_file_set_string (new_config, group, "xa.noenumerate", "true");
   g_key_file_set_string (new_config, group, "xa.prio", "0");
-  if (gpg_verify)
+  if (gpg_verify && collection_id == NULL)
     {
+      /* Don’t enable summary verification if a collection ID is set, as collection
```
Comment (pwithnall) [inline common/flatpak-dir.c]: Eek, yes, good catch.,

File: common/flatpak-dir.c [hunk 10]
```diff
@@ -8498,21 +9194,16 @@ flatpak_dir_fetch_ref_cache (FlatpakDir   *self,
...
-
-  summary = fetch_remote_summary_file (self, remote_name, NULL, cancellable, error);
-  if (summary == NULL)
-    return FALSE;
-
-  return flatpak_dir_parse_summary_for_ref (self, summary, ref,
-                                            download_size, installed_size,
-                                            metadata,
-                                            cancellable, error);
+  return flatpak_dir_parse_repo_metadata_for_ref (self, remote_name, ref,
```
Comment (pwithnall) [inline common/flatpak-dir.c]: Sure.

File: system-helper/flatpak-system-helper.c [hunk 11]
```diff
@@ -726,40 +727,62 @@ handle_update_remote (FlatpakSystemHelper *object,
...
       g_dbus_method_invocation_return_gerror  (invocation, error);
       return TRUE;
     }
+  summary_bytes = g_bytes_new_take (summary_data, summary_size);
 
+  if (collection_id == NULL)
+    {
+      if (!g_file_get_contents (arg_summary_sig_path, &summary_sig_data, &summary_sig_size, &error))
+        {
+          g_dbus_method_invocation_return_gerror  (invocation, error);
```
Comment (pwithnall) [inline system-helper/flatpak-system-helper.c]: Will fix in a separate commit.

File: common/flatpak-utils.c [hunk 12]
```diff
@@ -2552,69 +2552,119 @@ flatpak_variant_bsearch_str (GVariant   *array,
...
+          /* Must match type */
+          if (!g_str_has_prefix (cur, ref_prefix))
+            continue;
 
-      /* But only prefix of id */
-      if (!g_str_has_prefix (id_start + 1, parts_prefix))
-        continue;
+          /* Must match arch & branch */
+          if (!g_str_has_prefix (cur, ref_suffix))
+            continue;
```
Comment (pwithnall) [inline common/flatpak-utils.c]: Review comment from @alexlarsson on a plane: This reverts an existing fix (bad merge conflict resolution).

File: common/flatpak-dir.c [hunk 13]
```diff
@@ -2800,9 +3088,72 @@ flatpak_dir_pull_untrusted_local (FlatpakDir          *self,
...
+    {
+      gpg_result = ostree_repo_verify_commit_ext (src_repo, checksum, NULL, NULL, cancellable, error);
+      if (gpg_result == NULL)
+        return FALSE;
+
+      if (ostree_gpg_verify_result_count_valid (gpg_result) == 0)
+        return flatpak_fail (error, "GPG signatures found, but none are in trusted keyring");
+    }
+
+  g_clear_object (&gpg_result);
```
Comment (pwithnall) [inline common/flatpak-dir.c]: Review comment from @alexlarsson on a plane: This seems to pass `NULL` for the remote, which is a magic option that means we're matching against the keyrings from all remotes. That does not seem correct.

**Comment by pwithnall** [comment]
Other things which we discussed which I need to do/look at:
 * Talk to Richard about "lib: List dynamic remotes in flatpak_installation_list_remotes()"
 * Add collection ID to the manifest file in "builder: Add collection ID support to the flatpak builder"
 * gnome-software has its own parser for flatpak{ref,repo} files; needs to get collection-id support
 * Eventually add migration support for remote configurations to add collection-id and set gpg-verify-summary=false; add the code for that now, but don’t put the config in repos yet
 * Double-check OCI code paths; explicitly state that OCI+collection-id is unsupported
 * See if there’s a nice way to run the entire test suite with collection-id set and unset; see test-repo-system.sh

File: common/flatpak-utils.c (hunk 12 above)
Comment (pwithnall) [inline common/flatpak-utils.c]: Fixed.

File: common/flatpak-dir.c [hunk 14]
```diff
@@ -2800,9 +3088,72 @@ flatpak_dir_pull_untrusted_local (FlatpakDir          *self,
...
 
+  if (gpg_verify)
+    {
+      gpg_result = ostree_repo_verify_commit_ext (src_repo, checksum, NULL, NULL, cancellable, error);
+      if (gpg_result == NULL)
+        return FALSE;
+
+      if (ostree_gpg_verify_result_count_valid (gpg_result) == 0)
+        return flatpak_fail (error, "GPG signatures found, but none are in trusted keyring");
+    }
```
Comment (pwithnall) [inline common/flatpak-dir.c]: Review comment from @alexlarsson on a plane: This seems to pass `NULL` for the remote, which is a magic option that means we're matching against the keyrings from all remotes. That does not seem correct.

Comment (pwithnall) [inline common/flatpak-dir.c]: Fixed.

Comment (pwithnall) [inline common/flatpak-dir.c]: (To clarify things: `ostree_repo_verify_commit_ext()` *internally* passes `NULL` for the remote name. We should be using `ostree_repo_verify_commit_for_remote()` instead, I think.

**Comment by pwithnall** [comment]
> Other things which we discussed which I need to do/look at:
> 
>  - Talk to Richard about "lib: List dynamic remotes in flatpak_installation_list_remotes()"

I’ve got a WIP branch here which changes gnome-software to ignore non-static remotes for now:
https://github.com/pwithnall/gnome-software/tree/lan-and-usb

>  - Add collection ID to the manifest file in "builder: Add collection ID support to the flatpak builder"

Done. The `builder: Add collection ID support to the flatpak builder` commit has been updated to include this.

>  - gnome-software has its own parser for flatpak{ref,repo} files; needs to get collection-id support

I can’t find such a parser. The code around line 2700 of `gs-flatpak.c` calls `flatpak_installation_install_ref_file()` with a `flatpakref` file which it loaded from disk (but didn’t parse) itself. Looks like `gs_flatpak_app_install_source()` gets called for `flatpakrepo` files, though, and it needs modifying.

WIP branch (which will be submitted to bugzilla.gnome.org eventually) here: https://github.com/pwithnall/gnome-software/tree/lan-and-usb

>  - Eventually add migration support for remote configurations to add collection-id and set gpg-verify-summary=false; add the code for that now, but don’t put the config in repos yet

Added.

>  - Double-check OCI code paths; explicitly state that OCI+collection-id is unsupported

TODO

>  - See if there’s a nice way to run the entire test suite with collection-id set and unset; see test-repo-system.sh

TODO

File: common/flatpak-dir.c (hunk 1 above)
Comment (pwithnall) [inline common/flatpak-dir.c]: (Since fixed.)

File: common/flatpak-dir.c [hunk 15]
```diff
@@ -1853,69 +1867,181 @@ repo_pull_one_dir (OstreeRepo          *self,
...
+          force_disable_deltas = TRUE;
+        }
+
+      if (force_disable_deltas)
+        {
+          g_variant_builder_add (&find_builder, "{s@v}", "disable-static-deltas",
+                                 g_variant_new_variant (g_variant_new_boolean (TRUE)));
+          g_variant_builder_add (&pull_builder, "{s@v}", "disable-static-deltas",
+                                 g_variant_new_variant (g_variant_new_boolean (TRUE)));
+        }
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: This line is duplicated

File: common/flatpak-dir.c [hunk 16]
```diff
@@ -1853,69 +1867,181 @@ repo_pull_one_dir (OstreeRepo          *self,
...
   flags |= OSTREE_REPO_PULL_FLAGS_BAREUSERONLY_FILES;
 
-  g_variant_builder_init (&builder, G_VARIANT_TYPE ("a{sv}"));
+#ifdef FLATPAK_ENABLE_P2P
+  if (!ostree_repo_get_remote_option (self, remote_name, "collection-id", NULL,
+                                      &collection_id, NULL))
+    g_clear_pointer (&collection_id, g_free);
 
-  if (dirs_to_pull)
+  if (collection_id != NULL)
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: It would be nice if a lot of the variant builder code here with the regular pull. For instance all the options are the same.

File: common/flatpak-dir.c [hunk 17]
```diff
@@ -1853,69 +1867,181 @@ repo_pull_one_dir (OstreeRepo          *self,
...
+
+      if (progress != NULL)
+        update_freq = GPOINTER_TO_UINT (g_object_get_data (G_OBJECT (progress), "update-frequency"));
+      if (update_freq == 0)
+        update_freq = FLATPAK_DEFAULT_UPDATE_FREQUENCY;
+
+      g_variant_builder_add (&find_builder, "{s@v}", "update-frequency",
+                             g_variant_new_variant (g_variant_new_uint32 (update_freq)));
+      g_variant_builder_add (&pull_builder, "{s@v}", "update-frequency",
+                             g_variant_new_variant (g_variant_new_uint32 (update_freq)));
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: This like is duplicated

File: common/flatpak-dir.c (hunk 15 above)
Comment (pwithnall) [inline common/flatpak-dir.c]: Not entirely: the two copies add to different `GVariantBuilder` instances. I couldn’t think of a nice way to refactor that to not duplicate it.

File: common/flatpak-dir.c (hunk 17 above)
Comment (pwithnall) [inline common/flatpak-dir.c]: Not entirely: the two copies add to different `GVariantBuilder` instances. I couldn’t think of a nice way to refactor that to not duplicate it.

File: common/flatpak-dir.c (hunk 16 above)
Comment (pwithnall) [inline common/flatpak-dir.c]: Maybe there is some subsequent refactoring to do here. I left it as-is because it was easier, and also makes the review a bit easier than a more complex refactor.

Something to address once this branch is merged?

File: lib/flatpak-remote.h [hunk 18]
```diff
@@ -25,6 +25,15 @@
 #ifndef __FLATPAK_REMOTE_H__
 #define __FLATPAK_REMOTE_H__
 
+#ifdef FLATPAK_ENABLE_P2P
+typedef enum
```
Comment (alexlarsson) [inline lib/flatpak-remote.h]: Please don't make the libflatpak public headers ifdef:ed.
It should be completely safe to always have these available, as they are purely declarative and there is no way to create such remotes anyway. That means even if they have to change later (which i doubt) that doesn't really matter.

File: common/flatpak-dir.c (hunk 16 above)
Comment (pwithnall) [inline common/flatpak-dir.c]: I’ve got a local commit to refactor this; I’ll create a PR once this is merged (since github doesn’t support PRs which depend on other ones).

File: lib/flatpak-ref.h [hunk 19]
```diff
@@ -71,4 +71,8 @@ FLATPAK_EXTERN char *        flatpak_ref_format_ref (FlatpakRef *self);
 FLATPAK_EXTERN FlatpakRef *   flatpak_ref_parse (const char *ref,
                                                  GError    **error);
 
+#ifdef FLATPAK_ENABLE_P2P
+FLATPAK_EXTERN const char *   flatpak_ref_get_collection_id (FlatpakRef *self);
+#endif  /* FLATPAK_ENABLE_P2P */
```
Comment (alexlarsson) [inline lib/flatpak-ref.h]: Same here. Lets make this non-optional. Its just never going to return anything but NULL anyway.

File: lib/flatpak-remote.h (hunk 18 above)
Comment (pwithnall) [inline lib/flatpak-remote.h]: OK. I added the `#ifdef`s to avoid breaking public API later if they do have to change — but if you’re happy with that, I’m happy with removing that `#ifdef` mess. Do you want me to do that as an update to this PR, or as a follow-up PR?

File: lib/flatpak-ref.h (hunk 19 above)
Comment (pwithnall) [inline lib/flatpak-ref.h]: Sure. Same question about whether you want that done in this PR or in a follow-up.

File: common/flatpak-dir.c (hunk 16 above)
Comment (pwithnall) [inline common/flatpak-dir.c]: Actually, I just pushed it to this PR: https://github.com/flatpak/flatpak/pull/884/commits/4bf896e971aa8cc82489d642eba0886c84498f24

File: lib/flatpak-ref.h (hunk 19 above)
Comment (pwithnall) [inline lib/flatpak-ref.h]: Also, if we’re making the getter function unconditionally public, I assume you also want the `FlatpakRef::collection-id` property to be unconditionally public?

File: common/flatpak-dir.c [hunk 20]
```diff
@@ -8536,6 +8623,139 @@ flatpak_dir_fetch_remote_summary (FlatpakDir    *self,
...
+
+      g_debug ("Calling system helper: Deploy");
+      if (!flatpak_system_helper_call_deploy_sync (system_helper,
+                                                   child_repo_path ? child_repo_path : "",
+                                                   helper_flags, OSTREE_REPO_METADATA_REF, remote_name,
+                                                   (const char * const *) subpaths,
+                                                   installation ? installation : "",
+                                                   cancellable,
+                                                   error))
+        return FALSE;
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: I wonder if this is the right approach. In the previous system the summary cache was per-user, but now you push the cache to the global repo. This means you might get e.g. authentication dialogs on read-only operations.

I think we need to change this, making the ostree-metadata cache a per-user thing.

File: common/flatpak-dir.c (hunk 15 above)
Comment (pwithnall) [inline common/flatpak-dir.c]: (Fixed in https://github.com/flatpak/flatpak/pull/884/commits/4bf896e971aa8cc82489d642eba0886c84498f24.)

File: common/flatpak-dir.c (hunk 17 above)
Comment (pwithnall) [inline common/flatpak-dir.c]: (Fixed in https://github.com/flatpak/flatpak/pull/884/commits/4bf896e971aa8cc82489d642eba0886c84498f24.)

File: lib/flatpak-remote.h (hunk 18 above)
Comment (pwithnall) [inline lib/flatpak-remote.h]: Fixed and needing review in https://github.com/flatpak/flatpak/pull/884/commits/eddd2082919399204c0ab67308cc92bed8b15724.

File: lib/flatpak-ref.h (hunk 19 above)
Comment (pwithnall) [inline lib/flatpak-ref.h]: Fixed and needing review in https://github.com/flatpak/flatpak/pull/884/commits/eddd2082919399204c0ab67308cc92bed8b15724, including making `FlatpakRef::collection-id` public.

File: common/flatpak-dir.c (hunk 20 above)
Comment (pwithnall) [inline common/flatpak-dir.c]: Right. Is there a per-user repo already which we could pull into? I’m not 100% sure about how all the system-helper bits interact and where they all live.

Comment (pwithnall) [inline common/flatpak-dir.c]: We’re deferring this to a follow-up task.

**Comment by pwithnall** [comment]
I’ve just autosquashed in the final fixups. This now has all the review fixes squashed in, and is ready to merge.

**Comment by pwithnall** [comment]
I’ve filed three follow-up tasks for issues discussed above during the review:
 - https://github.com/flatpak/flatpak/issues/969
 - https://github.com/flatpak/flatpak/issues/970
 - https://github.com/flatpak/flatpak/issues/971
